File: android-java/app/src/main/java/com/oursky/skeleton/MainActivity.java
package com.oursky.skeleton;

import android.os.Bundle;
import android.widget.FrameLayout;
import android.support.v7.app.AppCompatActivity;

import com.bluelinelabs.conductor.Conductor;
import com.bluelinelabs.conductor.Router;
import com.bluelinelabs.conductor.RouterTransaction;
import com.facebook.drawee.backends.pipeline.Fresco;

import com.oursky.skeleton.client.Login;
import com.oursky.skeleton.redux.AppState;
import com.oursky.skeleton.redux.AppStateObservable;
import com.oursky.skeleton.redux.ClientState;
import com.oursky.skeleton.redux.ViewAction;
import com.oursky.skeleton.ui.LoginScreen;
import com.oursky.skeleton.ui.MainScreen;
import com.oursky.skeleton.ui.SplashScreen;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;

import static com.oursky.skeleton.MainApplication.store;

@SuppressWarnings({"ConstantConditions", "Convert2MethodRef"})
public class MainActivity extends AppCompatActivity {
    private boolean mStoreRetained = false;
    private boolean mInSplash;
    private boolean mShowingMain;
    private Router mRouter;
    private CompositeDisposable mSubscriptions = new CompositeDisposable();

    //region Lifecycle
    //---------------------------------------------------------------
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fresco.initialize(this);
        if (!mStoreRetained) {
            mStoreRetained = true;
            MainApplication.retainStore(this);
        }
        FrameLayout layout = new FrameLayout(this);
        setContentView(layout);
        mRouter = Conductor.attachRouter(this, layout, savedInstanceState);
        mInSplash = true;
        mShowingMain = false;
        if (AppConfig.SPLASH_DURATION == 0) {
            showAppContent();
        } else {
            mRouter.setRoot(RouterTransaction.with(new SplashScreen()));
        }
    }
    @Override
    protected void onStart() {
        // NOTE: We need to get store ready before super.onStart(),
        //       otherwise Conductor will re-create our view and cause NPE upon using store
        if (!mStoreRetained) {
            mStoreRetained = true;
            MainApplication.retainStore(this);
        }
        super.onStart();
    }
    @Override
    protected void onResume() {
        super.onResume();
        // subscribe to login state, we'll switch between LoginScreen and MainScreen
        mSubscriptions.add(AppStateObservable.create()
                .observeOn(AndroidSchedulers.mainThread())
                .map(mapLoginState)
                .distinctUntilChanged()
                .subscribe(consumeLoginState)
        );
        if (mInSplash) {
            // Show main screen after a delay, can also do things like login to server
            getWindow().getDecorView().postDelayed(() -> showAppContent(), AppConfig.SPLASH_DURATION);
        }
    }
    @Override
    protected void onPause() {
        super.onPause();
        mSubscriptions.clear();
        // TODO: We going background, release resource here
    }
    @Override
    protected void onStop() {
        super.onStop();
        if (mStoreRetained) {
            mStoreRetained = false;
            MainApplication.releaseStore(this);
        }
    }
    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
    //---------------------------------------------------------------
    //endregion

    //region UI Events
    //---------------------------------------------------------------
    @Override
    public void onBackPressed() {
        if (!mRouter.handleBack()) {
            super.onBackPressed();
        }
    }
    //---------------------------------------------------------------
    //endregion

    private void showAppContent() {
        if (mInSplash) {
            mInSplash = false;
            Login.Output login = store().getState().client().login.data;
            boolean logined = login != null && login.result == Login.Output.Result.Success;
            mShowingMain = logined;
            mRouter.replaceTopController(RouterTransaction.with(logined ? new MainScreen() : new LoginScreen()));
            ViewAction.setTitle(store(), "Hello World");
        }
    }

    //region Redux
    //---------------------------------------------------------------
    private final Function<AppState,ClientState.APIState<Login.Output>>
            mapLoginState = (state) -> state.client().login;
    private final Consumer<ClientState.APIState<Login.Output>> consumeLoginState = (mapped) -> {
        if (mInSplash) return;  // skip if in splash screen
        Login.Output login = mapped.data;
        boolean logined = login != null && login.result == Login.Output.Result.Success;
        if (mShowingMain != logined) {
            mShowingMain = logined;
            mRouter.replaceTopController(RouterTransaction.with(logined ? new MainScreen() : new LoginScreen()));
        }
    };
    //---------------------------------------------------------------
    //endregion
}

File: android-kotlin/app/src/main/java/com/oursky/skeleton/MainActivity.kt
package com.oursky.skeleton

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.widget.FrameLayout
import com.bluelinelabs.conductor.Conductor
import com.bluelinelabs.conductor.Router
import com.bluelinelabs.conductor.RouterTransaction
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import com.facebook.drawee.backends.pipeline.Fresco
import com.oursky.skeleton.redux.ClientStore
import com.oursky.skeleton.redux.ViewStore
import com.oursky.skeleton.ui.LoginScreen
import com.oursky.skeleton.ui.MainScreen
import com.oursky.skeleton.ui.SplashScreen
import com.oursky.skeleton.MainApplication.Companion.store

class MainActivity : AppCompatActivity() {
    private val mSubscriptions = CompositeDisposable()
    private var mRouter: Router? = null
    private var mStoreRetained: Boolean = false
    private var mInSplash: Boolean = false
    private var mShowingMain: Boolean = false

    //region Lifecycle
    //---------------------------------------------------------------
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Fresco.initialize(this)
        if (!mStoreRetained) {
            mStoreRetained = true
            MainApplication.retainStore(this)
        }
        val layout = FrameLayout(this)
        setContentView(layout)
        mRouter = Conductor.attachRouter(this, layout, savedInstanceState)
        mRouter?.setRoot(RouterTransaction.with(SplashScreen()))
        mInSplash = true
        mShowingMain = false
    }
    override fun onStart() {
        // NOTE: We need to get store ready before super.onStart(),
        //       otherwise Conductor will re-create our view and cause NPE upon using store
        if (!mStoreRetained) {
            mStoreRetained = true
            MainApplication.retainStore(this)
        }
        super.onStart()
    }
    override fun onResume() {
        super.onResume()
        mSubscriptions.add(store().observe(store().client)
                .distinctUntilChanged()
                .observeOn(AndroidSchedulers.mainThread())
                .map(mapLoginState)
                .subscribe(consumeLoginState)
        )
        if (mInSplash) {
            // Show main screen after a delay, can also do things like login to server
            window.decorView.postDelayed({ showAppContent() }, AppConfig.SPLASH_DURATION)
        }
    }
    override fun onPause() {
        super.onPause()
        mSubscriptions.clear()
        // TODO: We going background, release resource here
    }
    override fun onStop() {
        super.onStop()
        if (mStoreRetained) {
            mStoreRetained = false
            MainApplication.releaseStore(this)
        }
    }
    override fun onDestroy() {
        super.onDestroy()
    }
    //---------------------------------------------------------------
    //endregion

    //region UI Events
    //---------------------------------------------------------------
    override fun onBackPressed() {
        if (mRouter?.handleBack() != true) {
            super.onBackPressed()
        }
    }
    //---------------------------------------------------------------
    //endregion

    //region Redux
    //---------------------------------------------------------------
    private val mapLoginState = Function<ClientStore.State, Boolean> { state ->
        state.isLogined
    }
    private val consumeLoginState = Consumer<Boolean> { mapped ->
        if (mInSplash) return@Consumer // skip if in splash screen
        if (mShowingMain != mapped) {
            mShowingMain = mapped
            mRouter?.replaceTopController(RouterTransaction.with(if (mShowingMain) MainScreen() else LoginScreen()))
        }
    }
    //---------------------------------------------------------------
    //endregion

    private fun showAppContent() {
        if (mInSplash) {
            mInSplash = false
            mShowingMain = store().client.state.isLogined
            mRouter?.replaceTopController(RouterTransaction.with(if (mShowingMain) MainScreen() else LoginScreen()))
            store().dispatch(ViewStore.Action.SetTitle("Hello World"))
        }
    }
}
